services/src/main/java/org/keycloak/services/clientpolicy/executor/HolderOfKeyEnforceExecutor.java

```diff
@@ -0,0 +1,120 @@
+import org.keycloak.jose.jws.JWSInputException;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.util.MtlsHoKTokenUtil;
+import org.keycloak.util.JsonSerialization;
+
+import java.io.IOException;
+
+public class HolderOfKeyEnforceExecutor extends AbstractAugumentingClientRegistrationPolicyExecutor {
+
+    private final KeycloakSession session;
+    private final ComponentModel componentModel;
+
+    private boolean useMtlsHokToken;
+
+    public HolderOfKeyEnforceExecutor(KeycloakSession session, ComponentModel componentModel) {
+        super(session, componentModel);
+        this.session = session;
+        this.componentModel = componentModel;
+    }
+
+    @Override
+    public String getName() {
+        return componentModel.getName();
+    }
+
+    @Override
+    public String getProviderId() {
+        return componentModel.getProviderId();
+    }
+
+    @Override
+    protected void augment(ClientRepresentation rep) {
+        if (Boolean.parseBoolean(componentModel.getConfig().getFirst(AbstractAugumentingClientRegistrationPolicyExecutor.IS_AUGMENT))) {
+            OIDCAdvancedConfigWrapper.fromClientRepresentation(rep).setUseMtlsHoKToken(true);
+        }
+    }
+
+    @Override
+    protected void validate(ClientRepresentation rep) throws ClientPolicyException {
+        useMtlsHokToken = OIDCAdvancedConfigWrapper.fromClientRepresentation(rep).isUseMtlsHokToken();
+        if (!useMtlsHokToken) {
+            throw new ClientPolicyException(OAuthErrorException.INVALID_CLIENT_METADATA, "Invalid client metadata: MTLS token in disabled");
+        }
+    }
+
+    @Override
+    public void executeOnEvent(ClientPolicyContext context) throws ClientPolicyException {
+        super.executeOnEvent(context);
+        HttpRequest request = session.getContext().getContextObject(HttpRequest.class);
+
+        switch (context.getEvent()) {
+
+            case TOKEN_REQUEST:
+                if (useMtlsHokToken) {
+                    AccessToken.CertConf certConf = MtlsHoKTokenUtil.bindTokenWithClientCertificate(request, session);
+                    if (certConf == null) {
+                        throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Client Certification missing for MTLS HoK Token Binding");
+                    }
+                }
+                break;
+
+            case TOKEN_REFRESH:
+            case TOKEN_REVOKE:
+            case USERINFO_REQUEST:
+            case LOGOUT_REQUEST:
+                if (useMtlsHokToken) {
+                    String clientAssertion = request.getDecodedFormParameters().getFirst(OAuth2Constants.CLIENT_ASSERTION);
```

Comment:
@HryhoriiHevorkian "cliet_assertion" can be used for JWT based client authentication that has been defined as `private_key_jwt` of OpenID Connect 1.0 Section 9. It is not our target.